Improve compatibility of some WebRTC Stats collection

bc-lee opened the pull request:
In particular, ensure that candidate-pair stats are collected correctly.

queengooborg commented on custom/tests.yaml:
```diff
@@ -240,18 +260,25 @@ __resources:
-              var stats = new Map([...stats1, ...stats2]);
-              callback(stats);
+          getStats(rpc1, function() {
+            getStats(rpc2, function() {
+              setTimeout(function() {
+                getStats(rpc1, function(stats1) {
+                  getStats(rpc2, function(stats2) {
+                    var stats = new Map([...stats1, ...stats2]);
+                    callback(stats);
+                  });
+                });
+              }, 2000);
             });
           });
-        }, 2000);
+        }, 1000);
```
The comment above doesn't quite match up with what this code is actually doing.  The comment states that we're calling getStats() again if the results are empty, but we're not checking if they are, simply getting stats twice in a row regardless.  Is there a benefit to getting the stats twice vs. adding a longer delay?

bc-lee replied:
Sorry, the current comment is somewhat misleading. The correct one would be like this:

> Some stats are loaded later than others. Wait a bit after creating
> the connection, then call getStats. After that, call getStats once more
> since some stats will be created after the first call to getStats.

Do you think it's okay?

queengooborg replied:
That sounds good to me!

queengooborg commented on custom/tests.yaml:
```diff
@@ -3284,17 +3311,28 @@ api:
         return false;
       }
 
+      // The following function is defined to help obtain individual stats types
+      // and entries
+      function getStatByTypeAndEntry(type, entry) {
+        for (var i = 0; i < stats.length; i++) {
+          if (stats[i].type === type
+            && Object.prototype.hasOwnProperty.call(stats[i], entry)
+            && stats[i][entry] !== null) {
+            return stats[i];
+          }
+        }
+        return false;
+      }
+
```
What do you think about merging this with `getStatByType()` and changing the behavior depending on if the `entry` variable is present?

bc-lee replied:
Yeah, it is a good idea. I applied your suggestion and updated the PR.

queengooborg commented on custom/tests.yaml:
```diff
@@ -3284,17 +3311,28 @@ api:
+      // and entries
+      function getStatByTypeAndEntry(type, entry) {
+        for (var i = 0; i < stats.length; i++) {
+          if (stats[i].type === type
+            && Object.prototype.hasOwnProperty.call(stats[i], entry)
+            && stats[i][entry] !== null) {
```
I think we can simplify this by doing the following:

```suggestion
          if (stats[i].type === type && !!stats[i][entry]) {
```

bc-lee replied:
Thank you! I may write un-Idomatic JavaScript code because I don't use it every day.

queengooborg commented on custom/tests.yaml:
```diff
@@ -3284,17 +3311,28 @@ api:
-        return !!stat && 'availableOutgoingBitrate' in stat;
+        return !!getStatByTypeAndEntry('candidate-pair', 'availableOutgoingBitrate');
       type_candidate-pair.bytesDiscardedOnSend: |-
         var stat = getStatByType('candidate-pair');
         return !!stat && 'bytesDiscardedOnSend' in stat;
```
For consistency, we should probably use the `getStatByTypeAndEntry()` function everywhere (or the similar behavior on `getStatByType(type, entry)` if you'd like to go that route)!

queengooborg reviewed (changes requested):
Hey there, thank you so much for your PR to help improve these tests!  Web RTC is something that I am not very familiar with, so I am very glad to have someone more knowledgeable help fine-tune our tests!

I've just got a couple of suggestions, but this is otherwise LGTM!

bc-lee commented:
I used the following python script to regenerate yaml files consistently.

```python
#!/usr/bin/env python3

import sys
from typing import Dict

# third_party
import yaml

INPUT_FILE = "custom/tests.yaml"


def main():
  with open(INPUT_FILE, 'r') as f:
    data = yaml.safe_load(f)

  old_data = data['api']['RTCStatsReport']['__additional']
  keys = sorted(old_data.keys())
  new_data: Dict[str, str] = {}

  for key in keys:
    if not key.startswith("type_"):
      raise Exception("Invalid key: " + key)

    key_without_prefix = key[5:]
    if "." in key_without_prefix:
      type_name = key_without_prefix.split(".")[0]
      member_name = key_without_prefix.split(".")[1]
    else:
      type_name = key_without_prefix
      member_name = None

    old_content = old_data[key]
    if "XXX Need to wait" in old_content:
      new_data[key] = old_content
    elif member_name is None:
      new_data[key] = f"return !!getStatByType('{type_name}', null)"
    else:
      new_data[key] = f"return !!getStatByType('{type_name}', '{member_name}');"

  for key in keys:
    print(f"{key}: |-")
    data_for_print = new_data[key].replace('\n', '\n  ')
    print(f"  {data_for_print}")


if __name__ == '__main__':
  sys.exit(main())
```

bc-lee commented:
Thanks for the cleanups!

queengooborg commented:
Thank you for your hard work on this, @bc-lee!  I'm going to get this merged shortly so that it can be included in the next collector release!

Just a quick question: do you think that the second stats get call will obtain the `remote-inbound-rtp` and `remote-outbound-rtp` stats?  In other words, should we drop the `return {result: null, ...}` if `stats` is falsy on those stats types?

queengooborg reviewed (approved):
Once this is merged, I plan to include it in a collector release very soon so that we can get BCD updated ASAP!

bc-lee commented:
For answering your questions,
`remote-inbound-rtp` and `remote-outbound-rtp` stats may not be available at the beginning of the call since these stats are generated using RTCP packets sent by the remote peer.
So, we may need to wait for a while to get these stats.
2 seconds wait is usually enough for locally connected peers, but sometimes it may take longer for remote peers and it might introduce flaky results.
